### PR TITLE
Possibility to specify IPC path & RPC over HTTP support

### PR DESCRIPTION
**Pull request by tomusdrw**
Based on #393 

1. With this PR it would be possible to specify path to ipc socket (`--rpc` cli option)
2. There is also mock version of `net.Socket` which uses HTTP as a transport.

I would really love to get rid of that: https://github.com/ethereum/mist/compare/develop...tomusdrw:rpc?expand=1#diff-a31b48952900046cc9945315da84b939R22
It seems that the responses are stuck somewhere if the requests are sent synchronously (`setImmediate` or lower timeouts are also not sufficient). But the response is genererated by server and also quiting the application causes the response to be received and processed (sic!). Any hints?

### CODE REVIEW

**Comment by frozeman** [comment]
I will first merge @hiddentao PR and will then look at yours.

**Comment by frozeman** [comment]
Can you rebase your changes on develop?

**Comment by hiddentao** [comment]
Would be nice to get this one in.

File: modules/ipc/getIpcPath.js [hunk 1]
```diff
@@ -9,8 +9,12 @@ const log = require('../utils/logger').create('getIpcPath');
 
 module.exports = function() {
     var p = require('path');
-    var path = global.path.HOME;
+    var providedPath = global.rpcUri;
```
Comment (frozeman) [inline modules/ipc/getIpcPath.js]: why you create this variable here? Why not simply `if(global.rpcUri)` ?

File: modules/ipc/httpCompatSocket.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+Object.assign(HttpCompatSocket.prototype, {
+    call: null,
+    encoding: 'utf8',
+    writable: true, // socket is always writable
+
+    connect(event) {
+        this.call(JSON.stringify({
+            jsonrpc: "2.0",
+            id: 0,
+            method: "eth_syncing",
```
Comment (frozeman) [inline modules/ipc/httpCompatSocket.js]: better use `eth_accounts` for that, as `eth_syncing` is not available in older nodes.

**Comment by tomusdrw** [comment]
Thanks for comments. Updated the PR. Any hint what might be wrong with:
https://github.com/ethereum/mist/pull/420/files#diff-a31b48952900046cc9945315da84b939R22
(perhaps some better way to resolve this?)

**Comment by frozeman** [comment]
Try septime out with 0 or 1 instead of 50. But not sure what could cause this. I will have a look if i find time

**Comment by alexvandesande** [comment]
@hiddentao I can't assign you this issue, but I'll leave this to you. Can you please review, and solve the merge issues? If you consider it useful then I'll merge it

**Comment by tomusdrw** [comment]
Merged latest `develop` branch. Was trying lower values in `setTimeout` but the behaviour was inconsistent.

**Comment by hiddentao** [comment]
Sure I'll take a look at this. Actually my other PR changes a whole bunch of stuff so this might need some rewrites anwway. I hope to have that PR ready for review tomorrow.

**Comment by hiddentao** [comment]
I think I'd like to take these changes one a time, so first merge in #601 to enable custom IPC paths. And then look at the possibility of a non-IPC connection to geth, as enabled by the `HttpCompatSocket` in this PR.

**Comment by alexvandesande** [comment]
@hiddentao have your changes made this one redundant?

**Comment by hiddentao** [comment]
Not quite. This one also suggest enabling the use of HTTP instead of IPC for RPC purposes. That part is still not integrated into our code.

-- 
Ramesh Nair

Director
HiddenTao Ltd (UK company no. 6807289)

http://hiddentao.com
https://github.com/hiddentao
https://www.linkedin.com/in/hiddentao


On 2 June 2016 at 03:57:44, Alex Van de Sande (notifications@github.com) wrote:

@hiddentao have your changes made this one redundant?

**Comment by hiddentao** [comment]
Good idea would be to popup a warning message when user's use a HTTP connector, informing them that it's more risky than IPC.

**Comment by hiddentao** [comment]
Replaced by #871

**Pull request closed without merging**